src/dataman2/RowWriter.cpp

```diff
@@ -0,0 +1,383 @@
+    }
+    return cord_.size()  // data length
+           + offsetBytes * blockOffsets_.size()  // block offsets length
+           + nullFlags_.size()  // Number of bytes used for null flags
+           + verBytes  // version number length
+           + 1;  // Header
+}
+
+
+std::string RowWriter::encode() {
+    std::string encoded;
+
+    auto numFields = schema_->getNumFields();
+    if (numFields > 0) {
+        nullFlags_.resize(((numFields - 1) >> 3) + 1, 0);
+    }
+
+    // Reserve enough space so resize will not happen
+    encoded.reserve(sizeof(int64_t) * blockOffsets_.size()
+                    + cord_.size()
+                    + nullFlags_.size()
+                    + 11);
+    encodeTo(encoded);
+
+    return encoded;
+}
+
+
+void RowWriter::encodeTo(std::string& encoded) {
+    if (!schemaWriter_) {
+        // Padding the remaining fields. Could throw
+        operator<<(Skip(schema_->getNumFields() - colNum_));
+    }
+
+    // Header information
+    auto offsetBytes = calcOccupiedBytes(cord_.size());
+    char header = offsetBytes - 1;
+
+    SchemaVer ver = schema_->getVersion();
+    if (ver > 0) {
+        auto verBytes = calcOccupiedBytes(ver);
+        header |= verBytes << 5;
+        encoded.append(&header, 1);
+        // Schema version is stored in Little Endian
+        encoded.append(reinterpret_cast<char*>(&ver), verBytes);
+    } else {
+        encoded.append(&header, 1);
+    }
+
+    // Null flags follow the header byte
+    for (auto& f : nullFlags_) {
+        encoded.append(reinterpret_cast<char*>(&f), 1);
+    }
+
+    // Offsets follow the Null flags. They are persisted in Little Endian
+    for (auto offset : blockOffsets_) {
+        encoded.append(reinterpret_cast<char*>(&offset), offsetBytes);
+    }
+
+    cord_.appendTo(encoded);
+}
+
+
+Schema RowWriter::moveSchema() {
+    if (schemaWriter_) {
+        return schemaWriter_->moveSchema();
+    } else {
+        return Schema();
+    }
+}
+
+
+int64_t RowWriter::calcOccupiedBytes(uint64_t v) const noexcept {
+    int64_t bytes = 0;
+    do {
+        bytes++;
+        v >>= 8;
+    } while (v);
+
+    return bytes;
+}
+
+
+const cpp2::ValueType* RowWriter::getColumnType(
+        int64_t colIndex,
+        cpp2::SupportedType valueType) noexcept {
+    if (colIndex >= static_cast<int64_t>(schema_->getNumFields())) {
+        CHECK(!!schemaWriter_) << "SchemaWriter cannot be NULL";
+        if (!colType_) {
+            colType_.reset(new ColType(valueType));
+        }
+        return &(colType_->type_);
+    } else {
+        return &(schema_->getFieldType(colIndex));
+    }
+}
+
+
+void RowWriter::writeNullValue(int64_t colIndex, NullValue::NullType type) noexcept {
+    // set the Null flag
+    size_t byteIdx = colIndex >> 3;
+    nullFlags_.resize(byteIdx + 1, 0);
+    nullFlags_.back() |= ((uint8_t)0x80 >> (colIndex & 0x00000007));
+
+    // Output the Null value type
+    cord_ << static_cast<int8_t>(type);
```

Comment:
I think the `nullFlags_` is enough for indicate whether a value is null, so I guess the type is used for NaN and other special values?